<!-- FroboScout launch file -->

<launch>
	<!-- Frobit parameters --> 
	<param name="diff_steer_wheel_distance" value="0.215"/>
	<param name="ticks_per_meter_left" value="-562" />
	<param name="ticks_per_meter_right" value="-562" />

	<param name="robot_max_velocity" value="1.0"/>
	<param name="diff_steer_wheel_radius" value="0.051"/>
	<param name="diff_steer_wheel_ticks_per_rev" value="180"/>

	<node pkg="rosbag" name="rosbag_record" type="record" args="-a"/>

	<group ns="fmSensors">
		<node pkg="computer_load_monitor" name="load_monitor" type="computer_load_monitor_node" output="screen" >
			<param name="update_rate" value="5" />
		</node>

		<!-- interface to the Hokuyo lidar -->
		<param name="use_rep_117" value="true"/>
		<!--node name="hokuyo" pkg="hokuyo_node" type="hokuyo_node" output="screen" >
			<param name="port" value="/dev/ttyACM0"/>
			<param name="frame_id" type="string" value="/laser_link"/> 
			<param name="calibrate_time" value="true"/>
			<remap from="scan" to="/fmInformation/laser_msg"/>
			<param name="min_ang" type="double" value="-0.785" />
			<param name="max_ang" type="double" value="0.785" />
		 </node-->

		<!-- extract data from messages transmitted by the VectorNav VN-100 IMU -->
		<node pkg="vectornav_vn100" name="imu_node"  type="vectornav_vn100_node" output="screen">
			<param name="vectornav_vn100_sub" value="/fmData/imu_rx"/>
			<param name="imu_pub" value="/fmInformation/imu"/>
			<param name="frame_id" value="imu_link"/>
			<param name="use_enu" value="true"/>
			<param name="covariance_x" value="0.001"/>
			<param name="covariance_y" value="0.001"/>
			<param name="covariance_z" value="0.001"/>
		 </node>

		<!-- extract data from $GPGGA NMEA messages transmitted by the GPS -->
		<!--node pkg="gps_gpgga" name="gnss_gpgga" type="gps_gpgga_node" output="screen" respawn="true">
		    <param name="subscribe_topic_id" value="/fmData/gnss_rx" />
			<param name="publish_topic_id" value="/fmInformation/gpgga" />
		</node-->

		<!-- append Transverse Mercator projection coordinates to NMEA $GPGGA message -->
		<!--node pkg="gpgga_to_tranmerc" name="gpgga_to_tranmerc" type="gpgga_to_tranmerc_node" output="screen">
			<param name="gpgga_sub" value="/fmInformation/gpgga" />
			<param name="gpgga_tranmerc_pub" value="/fmInformation/gpgga_tranmerc" />
			<param name="transverse_mercator_a" value="6378137.0" />
			<param name="transverse_mercator_f" value="0.0033528106647474805" /> 
			<param name="transverse_mercator_false_easting" value="500000.0" />
			<param name="transverse_mercator_scale_factor" value="0.9996" />
			<param name="transverse_mercator_origin_latitude" value="0.0" />
			<param name="transverse_mercator_central_meridian" value="9.0" />
			<param name="transverse_mercator_false_northing" value="0.0" />
		</node-->
	</group>

	<group ns="fmProcessors">
		<!-- calculate odometry based on encoders and IMU yaw axis orientation -->
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/enc_left" />
			<param name="enc_right_sub" value="/fmInformation/enc_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_reset_sub" value="/fmInformation/odom_reset" />
			<!--param name="odom_pub" value="/fmKnowledge/odom" /-->
			<param name="odom_pub" value="/fmKnowledge/pose" />
			<param name="odom_estimate_frame" value="/odom_combined" />
			<param name="vehicle_frame" value="base_footprint" />
			<param name="encoder_output" value="relative" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="-z" />
		</node>

		<!-- estimate robot 2d pose -->
		<!--node pkg="pose_2d_estimator" name="pose_estimator" type="pose_2d_estimator_node.py" output="screen" >
			<param name="odom_sub" value="/fmKnowledge/odom" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="gga_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="pose_pub" value="/fmKnowledge/pose" />
			<param name="update_rate" value="50" />
			<param name="publish_relative_pose" type="boolean" value="false" />
			<param name="ekf_initial_guess_easting" value="0.0" />
			<param name="ekf_initial_guess_northing" value="0.0" />
			<param name="ekf_initial_guess_yaw" value="0.0" /> 
			<param name="odometry_distance_variance" value="0.000001" />
			<param name="odometry_angular_variance" value="0.00035" />
		</node-->

		<!-- node pkg="sdu_surveying_2014" name="survey_log" type="survey_log.py" output="screen" >
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="wptnav_status_sub" value="/fmInformation/wptnav_status" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="gpgga_tranmerc_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="lidar_sub" value="/fmSensors/laser_msg" />
		</node -->
	</group>

	<group ns="fmDecisionMakers">
		<node pkg="sdu_surveying_2014" name="mission" type="keyboard_mission_node.py" output="screen">
			<param name="keyboard_sub" value="/fmHMI/keyboard" />
			<param name="automode_pub" value="/fmDecision/automode" />
    		<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
            <param name="max_linear_velocity" value="0.8"/>
            <param name="max_angular_velocity" value="0.6"/>
            <param name="linear_velocity_step" value="0.1"/>
            <param name="angular_velocity_step" value="0.25"/>
		</node>

		<!--node pkg="sdu_surveying_2014" name="mission" type="wiimote_mission_node.py" output="screen">
			<param name="wiimote_sub" value="/fmLib/joy" />
			<param name="automode_pub" value="/fmDecision/automode" />
    		<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
            <param name="max_linear_velocity" value="0.7"/>
            <param name="max_angular_velocity" value="1.5"/>
            <param name="max_linear_acceleration" value="1.0"/>
            <param name="max_angular_acceleration" value="2.0"/>
            <param name="max_linear_deceleration" value="2.0"/>
            <param name="max_angular_deceleration" value="6.0"/>
		</node-->
	</group>

	<group ns="fmExecutors">
		<node pkg="sdu_surveying_2014" type="waypoint_navigation_node.py" name="wptnav" output="screen">
			<param name="automode_sub" value="/fmDecision/automode" />
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="joy_sub" value="/fmLib/joy" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
			<param name="implement_pub" value="/fmCommand/implement" />
			<param name="status_pub" value="/fmInformation/wptnav_status" />
			<param name="pid_pub" value="/fmInformation/wptnav_pid" />
			<param name="status_publish_interval" value="1" />
			<param name="pid_publish_interval" value="1" />
			<param name="print_debug_information" value="false" />

            <param name="max_linear_velocity" value="1.2"/>
            <param name="max_angular_velocity" value="0.70"/>

            <param name="drive_kp" value="1.6"/>
            <param name="drive_ki" value="0.0"/>
            <param name="drive_kd" value="0.0"/>
            <param name="drive_feed_forward" value="0.00"/>
            <param name="drive_max_output" value="0.50"/>

            <param name="turn_kp" value="1.8"/>
            <param name="turn_ki" value="0.0"/>
            <param name="turn_kd" value="0.0"/>
            <param name="turn_feed_forward" value="0.0"/>
            <param name="turn_max_output" value="1.0"/>

			<param name="wpt_default_mode" value="MCTE" />
			<param name="wpt_default_tolerance" value="0.20" />
			<param name="wpt_default_drive_velocity" value="0.4" />
			<param name="wpt_default_turn_velocity" value="0.6" />
			<param name="wpt_default_wait_after_arrival" value="2.0" />
			<param name="wpt_default_implement_command" value="0.0" />

			<param name="target_distance" value="0.4" />
			<param name="turn_start_at_heading_err" value="90.0" />
			<param name="turn_stop_at_heading_err" value="2.0" />
			<param name="ramp_drive_velocity_at_distance" value="1.5" />
			<param name="ramp_min_drive_velocity" value="0.15" />
			<param name="ramp_turn_velocity_at_angle" value="30.0" />
			<param name="ramp_min_turn_velocity" value="0.20" />
		</node>		
	</group>

	<group ns="fmControllers">
		<node pkg="differential_ifk" name="differential_driver" type="differential_ifk_node" output="screen" >
			<param name="hl_subscriber_topic" value="/fmCommand/cmd_vel"/>
			<param name="ll_publisher_topic_left" value="/fmSignal/cmd_vel_left"/>
			<param name="ll_publisher_topic_right" value="/fmSignal/cmd_vel_right"/>
		</node>	     
	</group>

	<group ns="fmLib">
		<node pkg="int_stamped_to_encoder" name="enc_conversion" type="encoder_converter.py">
			<param name="left_sub" value="/fmInformation/encoder_left" />
			<param name="right_sub" value="/fmInformation/encoder_right" />
			<param name="left_pub" value="/fmInformation/enc_left" />
			<param name="right_pub" value="/fmInformation/enc_right" />
		</node>

		<!-- Frobit interface node -->
		<node pkg="sdu_frobit" name="frobit_node" type="frobit_node">
			<param name="cmd_vel_left_sub" value="/fmSignal/cmd_vel_left"/>
			<param name="cmd_vel_right_sub" value="/fmSignal/cmd_vel_right"/>
			<param name="deadman_sub" value="/fmCommand/deadman"/>
			<param name="nmea_sub" value="/fmData/nmea_from_robot"/>
			<param name="nmea_pub" value="/fmData/nmea_to_robot"/>
			<param name="encoder_left_pub" value="/fmInformation/encoder_left"/>
			<param name="encoder_right_pub" value="/fmInformation/encoder_right"/>	
			<param name="castor_front" value="false"/>
			<param name="nmea_to_frobit_interval" value="0.05"/>          
			<param name="cmd_vel_timeout" value="1"/>          
        </node>

		<!-- Wiimote interface -->
		<!-- node pkg="wiimote" type="wiimote_node.py" name="wiimote" output="screen" respawn="true" -->
			<!-- param name="fixed_wiimote_address" value="00:1E:35:03:B8:81"/--> <!-- Kjeld SDU -->
			<!-- param name="fixed_wiimote_address" value="00:1F:32:B3:A2:F3"/ --> <!-- Kjeld -->
		<!--/node -->		

	 	<node pkg="keyboard_interface" type="keyboard_node" name="keyboard">
			<param name="keyboard_pub" value="/fmHMI/keyboard" />
		</node>

		<!-- Frobit -->
		<node pkg="serial_string" name="robot_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/robot_rx" />	 
			<param name="subscriber_topic" value="/fmData/robot_tx" />
			<param name="device" value="/dev/robot" />	
			<param name="baudrate" value="57600" />
			<param name="termination_character" value="10"/>
		</node>

		<node pkg="serial_nmea" name="robot_nmea" type="nmea_node" output="screen">
			<param name="str_to_msg_sub" value="/fmData/robot_rx" />   
			<param name="str_to_msg_pub" value="/fmData/nmea_from_robot" />
			<param name="msg_to_str_sub" value="/fmData/nmea_to_robot" />   
			<param name="msg_to_str_pub" value="/fmData/robot_tx" />           
			<param name="use_nmea_checksum" type="boolean" value="true" />
		</node>

		<!-- serial communication with the IMU -->
		<node pkg="serial_string" name="imu_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/imu_rx" />	 
			<param name="subscriber_topic" value="/fmData/imu_tx" />
			<param name="device" value="/dev/imu" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node> 	

		<!-- serial communication with the RTK-GNSS -->
		<!--node pkg="serial_string" name="gnss_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/gnss_rx" />	 
			<param name="subscriber_topic" value="/fmData/gnss_tx" />
			<param name="device" value="/dev/gps" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node-->

		<!--node pkg="real_time_analyze" name="rt_node" type="real_time_analyze_node.py" output="screen" >
			<param name="timing_pub" value="/fmInformation/rt_timing"/>
			<param name="device" value="/dev/rt_analyze" />	
        </node--> 	
	</group>
</launch>

